model-optimizer/extensions/back/MatMulNormalizer.py

```diff
@@ -142,3 +145,75 @@ def replace_pattern(graph: Graph, match: dict):
             src = port.get_source()
             port.get_connection().set_source(transpose_copy.out_port(0))
             src.connect(start_port)
+
+
+class SmartReshape_HC_Reshape_MatMul(BackReplacementPattern):
+    """
+    Relaxes hard-coded input of Reshape in such sub-graphs:
```

Comment:
nit: add a comment that the transformation works for MatMul with 2D input. 
Also, it will be a bit easier to read the code if the names in the diagram below correspond a bit to the names in the pattern.